http/events.go

```diff
@@ -69,14 +87,80 @@ func handleEventsSubscribeWebsocket(args eventSubscribeArgs) (websocket.StatusCo
 				logger.Warn("Could not serialize websocket event", "error", err)
 				return 0, "", err
 			}
-			err = args.conn.Write(ctx, messageType, messageBytes)
+			err = sub.conn.Write(ctx, messageType, messageBytes)
 			if err != nil {
 				return 0, "", err
 			}
 		}
 	}
 }
 
+// allowMessageCached checks that the message is allowed to received by the websocket.
+// It caches results for specific namespaces, data paths, and event types.
+func (sub *eventSubscriber) allowMessageCached(message *logical.EventReceived) bool {
+	// default to the subscribe endpoint if there is no specific data path
+	messageNs := strings.Trim(message.Namespace, "/")
+	dataPath := "sys/events/subscribe"
```

Comment:
I'm not sure whether we should instead fall back to the API path if the data path isn't available? Neither are guaranteed to be available though, so it's tricky to reason about what to do if neither are. It's also hard to say until we've got more event types implemented. Maybe to be on the safe side we should require root if no path is available, and if we find a natural extension in future we can add that?

Reply:
I remember we discussed what do with events with no paths in the RFC (like, say, a cluster change event), and landed on the policy should look like:

```
// for non-secret-related events, like cluster changes
path "ns1/*" {
  capabilities = ["subscribe"]
}
```

(presumably you'd also include only the event types allowed in such a policy as well).

But you're right, this feels a little weird. I don't think we can have an empty data path, since it wouldn't match a policy, and a bare wildcard seems like too much.

I think it is safer to do like you said, and if no path is available, then only allow root tokens, and we can try to figure out what a policy should look like for events with no paths.